### PR TITLE
Development/rex externals

### PR DESCRIPTION
**Pull request by Dyronix**

### CODE REVIEW

File: build/config/settings.json [hunk 1]
```diff
@@ -8,5 +8,5 @@
   "coverage_folder": "tests/coverage",
   "asan_folder": "tests/coverage",
   "ubsan_folder": "tests/ubsan",
-  "ninja_launcher": "build/scripts/rexpy/rexpy/ninja_launcher.py"
+  "ninja_launcher": "build/scripts/rex-py/rexpy/ninja_launcher.py"
```
Comment (RisingLiberty) [inline build/config/settings.json:11]: naming guideline I follow is that folders don't have '-' in there.
where did this come from?

File: _setup.py [hunk 2]
```diff
@@ -8,29 +8,23 @@
...
+  p.communicate()
 
-  # first let's make sure we're in the root directory
-  # we don't have access to rexpy yet, we'll have to manually
-  # check if we're in the root.
+def __main():
+  parser = argparse.ArgumentParser()
+  args, unknown = parser.parse_known_args()
 
   # let's assume if "build" and "source"are found, we're in the root
```
Comment (RisingLiberty) [inline _setup.py:29]: update this comment

File: _setup.py [hunk 3]
```diff
@@ -8,29 +8,23 @@
...
-  return False
+def __get_script_path():
+    return os.path.dirname(os.path.realpath(__file__))
 
-def install_rexpy(forceInstall):
-  # first set our working directory to rexpy.
-  # this way the package will be installed in {root}/build/scripts/rexpy
-  # this keeps the root clean.
+def __install_externals():
+  install_externals_path = os.path.join(__get_script_path(), os.path.join("build", os.path.join("scripts", "install_externals.py")))
```
Comment (RisingLiberty) [inline _setup.py:21]: this path should be wrapped in quotes

File: _setup.py [hunk 4]
```diff
@@ -40,39 +34,16 @@ def install_rexpy(forceInstall):
...
-  parser.add_argument("-update_rexpy", help="update rexpy, this is basically a reinstall", action="store_true")
-
-  args, unknown = parser.parse_known_args()
-
-  install_rexpy(args.update_rexpy)
-
-  # now that rexpy is installed, we can safely call the rest of the code
+  __install_externals()
 
+  # now that rexpy and rexstd is installed, we can safely call the rest of the code
```
Comment (RisingLiberty) [inline _setup.py:39]: update this comment

File: build/scripts/install_externals.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+class Host(Enum):
+    UNKNOWN = 0
+    GITLAB = 1
+    GITHUB = 2
+
+def __get_script_path():
+    return os.path.dirname(os.path.realpath(__file__))
+def __get_root_path():
+    root = __get_script_path()
+    root = os.path.join(root, os.path.join("..", ".."))
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:17]: look for the build folder in parent, don't use relative paths

File: build/scripts/install_externals.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    if "gitlab" in path:
+        return Host.GITLAB
+    elif "github" in path:
+        return Host.GITHUB
+    
+    print("Unknown host!")
+    return Host.UNKNOWN
+
+def __build_gitlab_path(baseUrl, name, tag):
+    url = os.path.join(baseUrl, "-")
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:31]: write a comment what the format of gitlab looks like

File: build/scripts/install_externals.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    url = os.path.join(url, "archive")
+    url = os.path.join(url, tag)
+    url = os.path.join(url, name+"-"+tag+".zip")
+    url = url.replace("\\", "/")
+    return url
+def __build_github_path(baseUrl, tag):
+    url = os.path.join(baseUrl, "archive")
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:38]: write a comment what the format of github looks like

File: build/scripts/install_externals.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    url = os.path.join(url, "refs")
+    url = os.path.join(url, "tags")
+    url = os.path.join(url, tag+".zip")
+    url = url.replace("\\", "/")
+    return url
+def __build_host_path(baseUrl, name, tag):
+    host = __get_host(baseUrl)
+    match host:
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:46]: match is only supported in python 3.10, a lot of people won't be able to run this

File: build/scripts/install_externals.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    root = __get_root_path()
+
+    external_url = external["url"]
+    external_name = external["name"]
+    external_tag = external["tag"]
+    external_store = external["storage"]
+    external_store = external_store.replace("~", root)
+
+    # if the external is already present we need to check if we need to redownload anything
+    should_download = False
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:145]: put this in a function

File: build/scripts/install_externals.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    externals_dir = os.path.join(external_store, external_name)
+    if os.path.exists(externals_dir):
+        print("External found: " + external_name + " validating version ...")
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:148]: also print where you found it

File: build/scripts/install_externals.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        os.chdir(externals_dir)
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:149]: I don't think you have to change the directory, keep using the absolute paths, it avoids a missing error where you put the cwd back to what it was

File: build/scripts/install_externals.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+            version_data = __load_json(version_file)           
+            if version_data == None:
+                print("Invalid version data found, redownloading external: " + external_name)
+                should_download = True              
+            if not version_data["tag"] == external_tag:
+                should_download = True
+            else:
+                print("External: " + external_name + " is up to date (" + external_name + " " + external_tag + ")")
+        else:
+            should_download = True
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:161]: print that you didn't find a version file

File: build/scripts/install_externals.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
+import os
```
Comment (RisingLiberty) [inline build/scripts/install_externals.py:1]: this entire script should be moved to rexpy, now I know this script also downloads rexpy so I'm not sure how we want to solve this.
it's possible a hardcoded rexpy url to download works and we donwload it uncompressed and then call into rexpy to download the rest

File: _setup.py [hunk 14]
```diff
@@ -40,39 +34,16 @@ def install_rexpy(forceInstall):
...
-if __name__ == "__main__":
-  parser = argparse.ArgumentParser()
-  parser.add_argument("-update_rexpy", help="update rexpy, this is basically a reinstall", action="store_true")
-
-  args, unknown = parser.parse_known_args()
-
-  install_rexpy(args.update_rexpy)
-
-  # now that rexpy is installed, we can safely call the rest of the code
+  __install_externals()
```
Comment (RisingLiberty) [inline _setup.py:37]: pip needs to rerun on rexpy if it's downloaded to addd it to the python libraries

**Review by RisingLiberty — CHANGES_REQUESTED** [review]

File: build/config/settings.json (hunk 1 above)
Comment (Dyronix) [reply build/config/settings.json:11]: RexStd was called rex_std, RexPy was called rexpy and glad & glm where called rex-glad and rex-glm I just picked one

Comment (Dyronix) [reply build/config/settings.json:11]: If files are not allowed to have no spaces I think folders should follow a similar guideline and I quite like the "-" between rex and the name of the library

Comment (RisingLiberty) [reply build/config/settings.json:11]: spaces are not recommened, I stuck with '_' because the source also has underscores

Comment (Dyronix) [reply build/config/settings.json:11]: I will change it to an "_"

File: build/scripts/install_externals.py (hunk 5 above)
Comment (Dyronix) [reply build/scripts/install_externals.py:17]: We will use rexpy for this which will be installed using pip

File: build/scripts/install_externals.py (hunk 13 above)
Comment (Dyronix) [reply build/scripts/install_externals.py:1]: We will use rexpy for this which will be installed using pip

**Pull request merged**